login-module/src/core.js

```diff
@@ -0,0 +1,77 @@
+import constants from './constants'
+import { WebAuth } from 'auth0-js'
+import Quiz from './quiz'
+import Enrollment from './enrollment'
+import Certificate from './certificate'
+
+export default class GraphAcademyCore {
+	constructor(options = {}) {
+		this.options = { ...constants.DEFAULT_OPTIONS, ...options }
+		const hasRequiredOptions = this.hasRequiredOptions(options)
+		if (!hasRequiredOptions) {
+			console.log(`required params missing - one of ${constants.REQUIRED_OPTIONS.join(', ')}`)
+			return
+		}
+		this.webAuth = new WebAuth({
+			clientID: 'hoNo6B00ckfAoFVzPTqzgBIJHFHDnHYu',
+			domain: 'login.neo4j.com',
+			redirectUri: `${window.location.origin}/accounts/login`,
+			audience: 'neo4j://accountinfo/',
+			scope: 'read:account-info openid email profile user_metadata',
+			responseType: 'token id_token'
+		})
+		this.certificate = new Certificate(this.options.trainingClassName, this.options.stage)
+		this.enrollment = new Enrollment(this.options.trainingClassName, this.options.stage)
+		this.quiz = new Quiz(this.options.trainingClassName, this.options.stage)
+	}
+
+	hasRequiredOptions(options) {
+		return constants.REQUIRED_OPTIONS.every(item => options[item])
+	}
+
+	async checkSession() {
+		const { options, webAuth } = this
+		return new Promise((resolve, reject) => {
```

Comment:
> I don't think you need to return Promise explicitly. If its async function it will always return a promise. If you are looking for a way to send reject, here is an approach -

Oh nice! I keep forgetting about that, thanks for the tip!

> But additionally what I would recommend simply return [err, result] and not use reject because in a situation the user is not logged in, it's not really an error.

Conceptually, we still have an expected result and an unexpected result. In Scala, this construct is an Either monad with a Right (authentication result) and a Left (error).
I don't mind using this construct instead of try/catch with await :+1:

Reply:
> I don't think you need to return Promise explicitly. If its async function it will always return a promise. If you are looking for a way to send reject, here is an approach -

It's not needed in the callback function, but I think I still need it to convert the callback style call to Auth0 to a Promise right?

Reply:
> But additionally what I would recommend simply return [err, result] and not use reject because in a situation the user is not logged in, it's not really an error.

Code becomes messy if we do that so I think we should definitely move the authenticated functions in another object so we can use the authentication context as a property.

Reply:
Interesting. I didn't know this came from Either monad concept. But I definitely feel this is a cleaner way to do it than using try/catch so let's use it.

Reply:
So totally on board with this idea of having authResult and accessToken as the class property.  So to confirm this would be the approach?

```js
const graphAcademy = new GraphAcademy();
const [err, result] = await graphAcademy.login();
if(err){
     // redirect to course landing page or do something
}
graphAcademy.getAccessToken();
graphAcademy.enrollStudentInClass('Shreyans', 'Gandhi');
```

and then the checkSession function in the core will be
```js
checkSession(){
    webAuth.checkSession((err, result) => {
        if(result){
            this.authResult = result;
        }
        return [err, result];
    })
}

getAccessToken() {
    return this.authResult ? this.authResult.accessToken : null;
}
```

This is how are you thinking of it?

Reply:
> So totally on board with this idea of having authResult and accessToken as the class property. So to confirm this would be the approach?

Almost:

```js
const graphAcademyCore = new GraphAcademyCore();
const [err, graphAcademyAuthenticated] = await graphAcademyCore.getAuthContext({});
if (err) {
     // redirect to course landing page or do something
}
graphAcademyAuthenticated.getAccessToken();
graphAcademyAuthenticated.enrollStudentInClass('Shreyans', 'Gandhi');
```

The value returned by `graphAcademy.login()` is either an error or an "authenticated object" on which you can call functions that needs an authenticated context.
As mentioned earlier, we don't want to rely on a mutable state that can be null/undefined. When using the "authenticated object" it's guaranteed that we have an access token.

> and then the checkSession function in the core will be

If we use the same example as above, we would have a `getAuthContext` function in core:

```js
class GraphAcademyAuthenticated {
  constructor(graphAcademy, authResult) {
    this.authResult = authResult
    this.graphAcademy = graphAcademy
  }
  getAccessToken() {
    return this.authResult.accessToken;
  }
  enrollStudentInClass(firstname, lastname) {
    // we don't have to check accessToken here
    const accessToken = getAccessToken()
  }
}

class GraphAcademyCore {
  getAuthContext (options) {
    webAuth.checkSession((err, result) => {
        if (err) {
          return [err, null]
        }
        if (result && result.accessToken) {
            return [null, new GraphAcademyAuthenticated(options, result)]
        }
        return [new Error('Invalid authentication result'), result]
    })
}
```